sql/analyze.go

```diff
@@ -0,0 +1,1036 @@
+// expressions of a list of expressions. The top-level list contains
+// disjunctions (a.k.a OR expressions). The second-level contains conjuntions
+// (a.k.a. AND expressions). For example:
+//
+//   (a AND b) OR c -> [[a, b], c]
+func analyzeExpr(e parser.Expr) []parser.Exprs {
+	e = simplifyExpr(e)
+	orExprs := splitOrExpr(e, nil)
+	results := make([]parser.Exprs, len(orExprs))
+	for i := range orExprs {
+		results[i] = splitAndExpr(orExprs[i], nil)
+	}
+	return results
+}
+
+// splitOrExpr flattens a tree of OR expressions returning all of the child
+// expressions as a list. Any non-OR expression is returned as a single element
+// in the list.
+//
+//   a OR b OR c OR d -> [a, b, c, d]
+func splitOrExpr(e parser.Expr, exprs parser.Exprs) parser.Exprs {
+	switch t := e.(type) {
+	case *parser.OrExpr:
+		return splitOrExpr(t.Right, splitOrExpr(t.Left, exprs))
+	}
+	return append(exprs, e)
+}
+
+// splitAndExpr flattens a tree of AND expressions returning all of the child
+// expressions as a list. Any non-AND expression is returned as a single
+// element in the list.
+//
+//   a AND b AND c AND d -> [a, b, c, d]
+func splitAndExpr(e parser.Expr, exprs parser.Exprs) parser.Exprs {
+	switch t := e.(type) {
+	case *parser.AndExpr:
+		return splitAndExpr(t.Right, splitAndExpr(t.Left, exprs))
+	}
+	return append(exprs, e)
+}
+
+// simplifyExpr transforms an expression such that it contains only expressions
+// involving qvalues that can be used for index selection. If an expression is
+// encountered that cannot be used for index selection (e.g. "func(val)") that
+// part of the expression tree is considered to evaluate to true, possibly
+// rendering the entire expression as true. Additionally, various
+// normalizations are performed on comparison expressions. For example:
+//
+//   (a < 1 AND a < 2)  -> (a < 1)
+//   (a < 1 AND a > 2)  -> false
+//   (a > 1 OR a < 2)   -> true
+//   (a > 1 OR func(b)) -> true
+//
+// Note that simplification is not normalization. Normalization as performed by
+// parser.NormalizeExpr returns an expression that is equivalent to the
+// original. Simplification can return an expression with parts of the
+// expression tree stripped out.
+func simplifyExpr(e parser.Expr) parser.Expr {
+	switch t := e.(type) {
+	case *parser.NotExpr:
+		return simplifyNotExpr(t)
+	case *parser.AndExpr:
+		return simplifyAndExpr(t)
+	case *parser.OrExpr:
+		return simplifyOrExpr(t)
+	case *parser.ComparisonExpr:
+		return simplifyComparisonExpr(t)
+	case *qvalue:
+		return e
+	}
+	// We don't know how to simplify expressions that fall through to here, so
+	// consider this part of the expression true.
+	return parser.DBool(true)
+}
+
+func simplifyNotExpr(n *parser.NotExpr) parser.Expr {
+	switch t := n.Expr.(type) {
+	case *parser.ComparisonExpr:
+		op := t.Operator
+		switch op {
+		case parser.EQ:
+			op = parser.NE
+		case parser.NE:
+			op = parser.EQ
+		case parser.GT:
+			op = parser.LT
```

Comment:
It sure is. More importantly, why didn't `checkEquivExpr` catch this...

Oh, it looks like `checkEquivExpr` was broken. Now fixed and it caught the errors here and some invalid equivalency checks in `TestSimplifyAndExprCheck`.